libcudacxx/include/cuda/std/detail/libcxx/include/__algorithm/copy_backward.h

```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___ALGORITHM_COPY_BACKWARD_H
+#define _LIBCUDACXX___ALGORITHM_COPY_BACKWARD_H
+
+#ifndef __cuda_std__
+#  include <__config>
+#endif // __cuda_std__
+
+#include "../__algorithm/copy.h"
+#include "../__algorithm/unwrap_iter.h"
+#include "../__type_traits/enable_if.h"
+#include "../__type_traits/is_same.h"
+#include "../__type_traits/is_trivially_copy_assignable.h"
+#include "../__type_traits/remove_const.h"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <class _BidirectionalIterator, class _OutputIterator>
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 _OutputIterator
+__copy_backward(_BidirectionalIterator __first, _BidirectionalIterator __last, _OutputIterator __result)
+{
+  while (__first != __last)
+  {
+    *--__result = *--__last;
+  }
+  return __result;
+}
+
+template <class _Tp,
+          class _Up,
+          __enable_if_t<_LIBCUDACXX_TRAIT(is_same, __remove_const_t<_Tp>, _Up), int> = 0,
+          __enable_if_t<_LIBCUDACXX_TRAIT(is_trivially_copy_assignable, _Up), int>   = 0>
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX17 _Up*
+__copy_backward(_Tp* __first, _Tp* __last, _Up* __result)
+{
+  const size_t __n = static_cast<size_t>(__last - __first);
+  if (__n > 0)
+  {
+    __result -= __n;
+    _CUDA_VSTD::__copy(__first, __last, __result);
```

Comment:
critical: although this works on host, it leads to wrong results on device (see remark on copy implementation):
```cpp
  int a[] = { 1, 2, 3, 4, 5 };
  int *begin = a;
  int *end = a + 3;
  int *result = a + 5;
  
  cuda::std::copy_backward(begin, end, result);

  for (int v: a) {
    std::printf("%d ", v);
  }
  std::printf("\n");
```
Sinde `d_last` doesn't overlap `[first, last)`, `copy_backward` can be used. 
Expected output is `1 2 1 2 3`, but when the algorithm is executed on device, it's `1 2 1 2 1`.

Reply:
Thanks a lot I have fixed the implementation and added tests to ensure that we properly copy the right elements